### PR TITLE
feature/324 | skip non PR notifications

### LINKED ISSUE #324: Branches and PRs overwrite each other statuses in Bitbucket allowing code to be merged before it is fully tested
### Your checklist for this issue

- [x] Jenkins version

- [x] Plugin version

- [ ] Bitbucket cloud

- [x] Bitbucket server and version

### Description

- Jenkins version: 2.235.1 (The Jenkins version is irrelevant)

- Plugin version: Any version >= 2.4.3 (we are currently using 2.8.1-rc646.71e4d09c0b37 as a result of this issue: https://github.com/jenkinsci/bitbucket-branch-source-plugin/issues/287)

- Bitbucket server and version: Server version 7.1.1 (The Bitbucket version is irrelevant)

- Issue: When a feature branch and a pull request branch are both running at the same time, whichever one finishes first causes the status in bitbucket to be updated, this will then be overwritten by the other job when it finishes. Therefore, failures may go undiagnosed either because they are overwritten by the second job, or be ignored because the code is merged before all branches have finished building.

### TLDR
The following issue occurs when Jenkins is configured to use the discover branch strategy "Exclude branches that are also filed as PRs"

When a user pushes code to a feature branch the corresponding job is triggered in Jenkins. If the user then creates a pull request in BitBucket whilst this branch is building it will cause another build to get triggerd in Jenkins. All of this is expected behaviour.

However, the "build status" of the PR in Bitbucket only shows 1 build. It shows the status of whichever build updated it last e.g. it will show the status of the feature branch until the PR is triggered, then it wil show the status of the PR, then, when the feature branch completes (but the PR is still building) it will show the 'completed' status of the feature branch, despite the fact there is still a PR job running.

This behaviour means that for those of us who run extended sets of tests for PRs there is a window between the feature branch completing and the PR branch completing where the status in Bitbucket is shown as "passed", when in reality there is still the possibility of a failure being reported by the PR job.

I have tracked this down to the following PR: https://github.com/jenkinsci/bitbucket-branch-source-plugin/pull/172 which was to address the following Jenkins ticket: https://issues.jenkins-ci.org/browse/JENKINS-45643
This was subsequently released in version 2.4.3 of this plugin and has been there ever since.

Now, having been a victim of JENKINS-45643 in the past myself (although I've always worked around it by either deleting the PR and triggering the feature branch manually, or hitting the Bitbucket REST API to override the status of the failed/pending job) I understand why the approach in PR-172 was taken. However, it overlooked the need for the status of the PR build to take precedence over all over branches.

This creates a difficult situation on how to proceed as we don't want to introduce a regression to the plugin by removing this functionality. That said, the position we're currently in is that users of the plugin are potentially being misled by the status of jobs appearing in Bitbucket.

One 'workaround' that I have considered is to stop using the "Exclude branches that are also filed as PRs" and just use "All branches" but that will massively duplicate the number of builds happening in Jenkins.

### Proposed solution

My proposal (if it's possible to implement) would be to add a tickbox in the configuration of the plugin (much like the recent additions of "Call Can Merge" and "Call changes api") which would allow users of the plugin to "opt in" to using the older style of naming where feature branches and PR branches show uniquely in Bitbucket, perhaps something like "Branches and PRs use different Bitbucket identifiers".

I believe this would also resolve this issue: https://github.com/jenkinsci/bitbucket-branch-source-plugin/issues/298

### PR DESCRIPTION
**Pull request by sams-gleb**
Adding a checkbox to avoid overwriting of _PR_ status by the _branch_ build status. Should fix #324 . We had similar problems in our pipelines where branch build can be finished in the middle of running PR build thus allowing to merge un-tested changes into master

### Your checklist for this pull request

- [x ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side) and not your master branch!
- [ x] Ensure that the pull request title represents the desired changelog entry
- [x ] Please describe what you did
- [x ] Link to relevant issues in GitHub or in [Jenkins JIRA](https://issues.jenkins-ci.org)
- [x ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Did you provide a test-case? That demonstrates feature works or fixes the issue.

### CODE REVIEW

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketBuildStatusNotifications.java [hunk 1]
```diff
@@ -199,6 +204,15 @@ private static void sendNotifications(BitbucketSCMSource source, Run<?, ?> build
             listener.getLogger().println("[Bitbucket] Notifying commit build result");
             key = getBuildKey(build, r.getHead().getName(), shareBuildKeyBetweenBranchAndPR);
             bitbucket = source.buildBitbucketClient();
+            if (sourceContext.doNotOverridePrBuildStatuses()) {
+                BitbucketBuildStatuses buildStatuses = bitbucket.getBuildStatus(hash);
+                for (BitbucketBuildStatus bs : buildStatuses.getValues()) {
```
Comment (KalleOlaviNiemitalo) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketBuildStatusNotifications.java:209]: Looks like this will blow up if getBuildStatus returns null.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketBuildStatusNotifications.java [hunk 2]
```diff
@@ -157,7 +161,7 @@ private static void createStatus(@NonNull Run<?, ?> build, @NonNull TaskListener
             statusDescription = StringUtils.defaultIfBlank(buildDescription, "The build is in progress...");
             state = INPROGRESS_STATE;
         }
-        status = new BitbucketBuildStatus(hash, statusDescription, state, url, key, name);
+        status = new BitbucketBuildStatus(hash, statusDescription, state, url, DigestUtils.md5Hex(key), name);
```
Comment (KalleOlaviNiemitalo) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketBuildStatusNotifications.java:164]: The added md5Hex call looked surprising but I now see you just moved it from the BitbucketBuildStatus constructor.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java [hunk 3]
```diff
@@ -648,6 +649,11 @@ public void postBuildStatus(@NonNull BitbucketBuildStatus status) throws IOExcep
         postRequest(url, JsonParser.toJson(status));
     }
 
+    @Override
+    public BitbucketBuildStatuses getBuildStatus(@NonNull String hash) throws IOException, InterruptedException {
+        return null;
+    }
+
```
Comment (KalleOlaviNiemitalo) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java:656]: And here it does return null, causing an exception in the caller. This should preferably be implemented, or at least the caller should check for null.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketBuildStatusNotifications.java [hunk 4]
```diff
@@ -199,6 +204,15 @@ private static void sendNotifications(BitbucketSCMSource source, Run<?, ?> build
             listener.getLogger().println("[Bitbucket] Notifying commit build result");
             key = getBuildKey(build, r.getHead().getName(), shareBuildKeyBetweenBranchAndPR);
             bitbucket = source.buildBitbucketClient();
+            if (sourceContext.doNotOverridePrBuildStatuses()) {
```
Comment (KalleOlaviNiemitalo) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketBuildStatusNotifications.java:207]: Should this have a special case for reporting the status of a pull request build?

Like, there is a pull request and it's first built once and the build status goes to Bitbucket all right. Then a user of Jenkins replays the build — there is no new commit. When the second build finishes, does the plugin decide not to report the status to Bitbucket because that commit already has one build status from a pull request — even though it's from the same pull request? It seems to me that the previous build status should be overwritten in that case.

File: src/main/resources/com/cloudbees/jenkins/plugins/bitbucket/BitbucketBuildStatusNotificationsTrait/config.jelly [hunk 5]
```diff
@@ -3,4 +3,7 @@
   <f:entry title="${%Communicate Unstable builds to Bitbucket as Successful}" field="sendSuccessNotificationForUnstableBuild">
     <f:checkbox/>
   </f:entry>
+   <f:entry title="${%Do not override PR build statuses}" field="doNotOverridePrBuildStatuses">
+    <f:checkbox/>
+  </f:entry>
```
Comment (KalleOlaviNiemitalo) [inline src/main/resources/com/cloudbees/jenkins/plugins/bitbucket/BitbucketBuildStatusNotificationsTrait/config.jelly:8]: I would like a longer help text that describes what this setting does and why one would want to use it. Yes the information is in <https://github.com/jenkinsci/bitbucket-branch-source-plugin/issues/324>, but it's nicer to have it in Jenkins as well.

**Review by KalleOlaviNiemitalo — CHANGES_REQUESTED** [review]

**Comment by KalleOlaviNiemitalo** [comment]
Overall though, I feel the better fix would be to (1) abort the branch build as soon as Jenkins finds that there is a PR from that branch as well and (2) delete the INPROGRESS build status from Bitbucket when the build is aborted that way.

**Comment by sams-gleb** [comment]
> Overall though, I feel the better fix would be to (1) abort the branch build as soon as Jenkins finds that there is a PR from that branch as well and (2) delete the INPROGRESS build status from Bitbucket when the build is aborted that way.

Thanks for the review! Makes sense, I will look when I have time and mb will close this PR

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketBuildStatusNotifications.java (hunk 4 above)
Comment (sams-gleb) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketBuildStatusNotifications.java:207]: for this case it shouldn't be a problem as notification for PRs being processed in the previous `if` statement, this is for "branches" only, you could check line 198

**Comment by sams-gleb** [comment]
> Overall though, I feel the better fix would be to (1) abort the branch build as soon as Jenkins finds that there is a PR from that branch as well and (2) delete the INPROGRESS build status from Bitbucket when the build is aborted that way.

I took another quick look.. The main use case here is when we create a PR and still having a running branch pipeline which could overwrite PR status thus non-tested changes could be merged. Branches and PRs do share key so it might be not so simple to delete only in progress branch status, in many cases it might be already in progress status of PR

File: src/main/resources/com/cloudbees/jenkins/plugins/bitbucket/BitbucketBuildStatusNotificationsTrait/config.jelly (hunk 5 above)
Comment (KalleOlaviNiemitalo) [reply src/main/resources/com/cloudbees/jenkins/plugins/bitbucket/BitbucketBuildStatusNotificationsTrait/config.jelly]: I meant, in `src/main/resources/com/cloudbees/jenkins/plugins/bitbucket/BitbucketBuildStatusNotificationsTrait/help-doNotOverridePrBuildStatuses.html`.

Comment (sams-gleb) [reply src/main/resources/com/cloudbees/jenkins/plugins/bitbucket/BitbucketBuildStatusNotificationsTrait/config.jelly]: thanks, added

**Comment by alexanderrtaylor** [comment]
Those failing builds are not legit. It appears that the filename has gotten too long due to the length of the git commit hash when checking out inside of windows.

That being said, I am not sure this feature is all that useful @sams-gleb because what about things like main/versioned branches that you do want to know build status on?

To me instead of a checkbox that ignores anything but PR build status, I think it should be a regex which would let you ignore status notifications from anything not-desired. To me that is a more useful feature and also you will most likely have to add a test before this gets approved and merged in. Because this is something super useful to me I am going to try and help make some code changes to see if I can get it working

**Comment by nfalco79** [comment]
Superseded by #930
